experimental/dds/tree2/src/feature-libraries/sequence-field/utils.ts

```diff
@@ -38,44 +34,41 @@ import {
 } from "./format";
 import { MarkListFactory } from "./markListFactory";
 import { isMoveMark, MoveEffectTable } from "./moveEffectTable";
-import { GenerativeMark, TransientMark, EmptyInputCellMark, DetachedCellMark } from "./helperTypes";
+import { TransientMark, EmptyInputCellMark, DetachedCellMark } from "./helperTypes";
 
 export function isEmpty<T>(change: Changeset<T>): boolean {
 	return change.length === 0;
 }
 
-export function isNewAttach<TNodeChange>(mark: Mark<TNodeChange>): mark is NewAttach<TNodeChange> {
-	return mark.type === "Insert" || mark.type === "MoveIn";
+export function isNewAttach<TNodeChange>(mark: Mark<TNodeChange>, revision?: RevisionTag): boolean {
+	return (
+		isAttach(mark) &&
+		mark.cellId !== undefined &&
+		(mark.revision ?? revision) === (mark.cellId.revision ?? revision)
+	);
 }
 
-export function isGenerativeMark<TNodeChange>(
-	mark: Mark<TNodeChange>,
-): mark is GenerativeMark<TNodeChange> {
-	return mark.type === "Insert" || mark.type === "Revive";
+export function isInsert<TNodeChange>(mark: Mark<TNodeChange>): mark is Insert<TNodeChange> {
+	return mark.type === "Insert";
 }
 
 export function isAttach<TNodeChange>(mark: Mark<TNodeChange>): mark is Attach<TNodeChange> {
-	return isNewAttach(mark) || isReattach(mark);
+	return mark.type === "Insert" || mark.type === "MoveIn";
 }
 
-export function isReattach<TNodeChange>(mark: Mark<TNodeChange>): mark is Reattach<TNodeChange> {
-	return mark.type === "Revive" || mark.type === "ReturnTo";
+export function isReattach<TNodeChange>(mark: Mark<TNodeChange>): boolean {
+	return isAttach(mark) && !isNewAttach(mark);
 }
 
 export function isActiveReattach<TNodeChange>(
 	mark: Mark<TNodeChange>,
-): mark is Reattach<TNodeChange> & { conflictsWith?: undefined } {
-	// No need to check Reattach.lastDeletedBy because it can only be set if the mark is conflicted
-	return isReattach(mark) && !isConflictedReattach(mark);
+): mark is Attach<TNodeChange> {
```

Comment:
Unfortunately, I don't think this works, as it will cause the compiler to infer that a new attach or inactive reattach is not an attach. It would be nice if there were a way to make one-way type guards.

Reply:
Yeah, we need to make a PR that fixes all these things.
I'll merge and do that.